api/admin/announcement_list_validator.py

```diff
@@ -0,0 +1,155 @@
+import datetime
+import json
+import uuid
+
+from flask_babel import lazy_gettext as _
+
+from api.admin.validator import Validator
+
+from core.util.problem_detail import ProblemDetail
+from core.problem_details import *
+
+
+class AnnouncementListValidator(Validator):
+
+    DATE_FORMAT = '%Y-%m-%d'
+
+    def __init__(self, maximum_announcements=3, minimum_announcement_length=15,
+                 maximum_announcement_length=350, default_duration_days=60):
+        super(AnnouncementListValidator, self).__init__()
+        self.maximum_announcements = maximum_announcements
+        self.minimum_announcement_length = minimum_announcement_length
+        self.maximum_announcement_length = maximum_announcement_length
+        self.default_duration_days = default_duration_days
+
+    def validate_announcements(self, announcements):
+        validated_announcements = []
+        bad_format = INVALID_INPUT.detailed(
+            _("Invalid announcement list format: %(announcements)r",
+              announcements=announcements)
+        )
+        if isinstance(announcements, basestring):
+            try:
+                announcements = json.loads(announcements)
+            except ValueError:
+                return bad_format
+        if not isinstance(announcements, list):
+            return bad_format
+        if len(announcements) > self.maximum_announcements:
+            return INVALID_INPUT.detailed(
+                _("Too many announcements: maximum is %(maximum)d",
+                  maximum=self.maximum_announcements)
+            )
+            
+        seen_ids = set()
+        for announcement in announcements:
+            validated = self.validate_announcement(announcement)
+            if isinstance(validated, ProblemDetail):
+                return validated
+            id = validated['id']
+            if id in seen_ids:
+                return INVALID_INPUT.detailed(_("Duplicate announcement ID: %s" % id))
+            seen_ids.add(id)
+            validated_announcements.append(validated)
+        return validated_announcements
+
+    def validate_announcement(self, announcement):
+        validated = dict()
+        if not isinstance(announcement, dict):
+            return INVALID_INPUT.detailed(
+                _("Invalid announcement format: %(announcement)r", announcement=announcement)
+            )
+
+        validated['id'] = announcement.get('id', unicode(uuid.uuid4()))
+
+        for required_field in ('start', 'content'):
```

Comment:
I don't think "start" is required; if the admin leaves the start date field blank, it will automatically be set to today's date.  (If the end date field is blank, it will be set to two months after the start date.  So it doesn't make sense to treat only one of them as required...)